# Authors and Citation

## Authors

- **Ilya Korsunsky**. Maintainer, author.
  [](https://orcid.org/0000-0003-4848-3948)

- **Daniel Stein**. Author. [](https://orcid.org/0000-0002-9442-2604)

## Citation

Source:
[`inst/CITATION`](https://github.com/korsunskylab/tessera/blob/main/inst/CITATION)

Stein DJ, Tran M, Korsunsky I (2025). “Accurate tiling of spatial
single-cell data with Tessera.” *bioRxiv*.
[doi:10.1101/2025.01.17.633630](https://doi.org/10.1101/2025.01.17.633630),
https://www.biorxiv.org/content/early/2025/01/22/2025.01.17.633630.full.pdf,
<https://www.biorxiv.org/content/early/2025/01/22/2025.01.17.633630>.

    @Article{,
      author = {Daniel Jiang Stein and Miles Tran and Ilya Korsunsky},
      title = {Accurate tiling of spatial single-cell data with Tessera},
      year = {2025},
      doi = {10.1101/2025.01.17.633630},
      publisher = {Cold Spring Harbor Laboratory},
      abstract = {Single-cell spatial transcriptomics reveals how cells organize in healthy and diseased tissues. From these data, tissue segmentation analysis defines discrete compartments that organize cells into functional multicellular units. Existing methods for automated tissue segmentation rely on spatial smoothing to define spatially coherent regions but often blur the boundaries between adjacent tissue compartments. We describe Tessera, an algorithm that approaches tissue segmentation through a novel approach, dividing the tissue into small multicellular tiles whose edges track with natural tissue boundaries. Tessera achieves this by incorporating successful tools from edge-preserving smoothing, topological data analysis, and morphology-aware agglomerative spatial clustering. We show that Tessera identifies a range of known anatomical structures, in healthy mouse brain and human lymph nodes, and novel disease-associated niches, in human brain and in lung cancer. Tessera is a general-purpose tool that returns spatially coherent spatial structures with accurate boundaries across a range of spatial transcriptomics and proteomics technologies.Competing Interest StatementThe authors have declared no competing interest.},
      url = {https://www.biorxiv.org/content/early/2025/01/22/2025.01.17.633630},
      eprint = {https://www.biorxiv.org/content/early/2025/01/22/2025.01.17.633630.full.pdf},
      journal = {bioRxiv},
    }
